File: functions/shared/race_eligibility.py
"""
Race Eligibility Calculation Engine
Determines which races a crew is eligible for based on age and gender composition
"""
from datetime import datetime, date
from typing import List, Dict, Any, Optional


def calculate_age(date_of_birth: str, reference_date: Optional[date] = None) -> int:
    """
    Calculate age based on date of birth
    
    Args:
        date_of_birth: Date string in YYYY-MM-DD format
        reference_date: Reference date for age calculation (defaults to today)
    
    Returns:
        Age in years
    """
    if reference_date is None:
        reference_date = date.today()
    
    birth_date = datetime.strptime(date_of_birth, '%Y-%m-%d').date()
    age = reference_date.year - birth_date.year
    
    # Adjust if birthday hasn't occurred this year
    if reference_date.month < birth_date.month or \
       (reference_date.month == birth_date.month and reference_date.day < birth_date.day):
        age -= 1
    
    return age


def get_age_category(age: int) -> str:
    """
    Determine age category based on age
    
    Args:
        age: Age in years
    
    Returns:
        Age category string (j16, j18, senior, master)
    """
    if age <= 16:
        return "j16"
    elif age <= 18:
        return "j18"
    elif age < 27:
        return "senior"
    else:
        return "master"


def get_master_category(avg_age: float) -> str:
    """
    Determine master category letter based on average age of crew
    
    Args:
        avg_age: Average age of crew members
    
    Returns:
        Master category letter (A, B, C, D, E, F, G, H)
    """
    if avg_age < 36:
        return "A"  # 27-35
    elif avg_age < 43:
        return "B"  # 36-42
    elif avg_age < 50:
        return "C"  # 43-49
    elif avg_age < 55:
        return "D"  # 50-54
    elif avg_age < 60:
        return "E"  # 55-59
    elif avg_age < 65:
        return "F"  # 60-64
    elif avg_age < 70:
        return "G"  # 65-69
    else:
        return "H"  # 70+


def analyze_crew_composition(crew_members: List[Dict[str, Any]]) -> Dict[str, Any]:
    """
    Analyze crew composition for race eligibility
    
    Args:
        crew_members: List of crew member objects with date_of_birth and gender
    
    Returns:
        Dictionary with crew composition analysis
    """
    if not crew_members:
        return {
            'crew_size': 0,
            'genders': [],
            'ages': [],
            'age_categories': [],
            'gender_category': None,
            'age_category': None,
            'eligible_boat_types': []
        }
    
    # Calculate ages and determine categories
    ages = []
    genders = []
    age_categories = []
    
    for member in crew_members:
        age = calculate_age(member['date_of_birth'])
        ages.append(age)
        genders.append(member['gender'])
        age_categories.append(get_age_category(age))
    
    # Determine gender category
    unique_genders = set(genders)
    if len(unique_genders) == 1:
        gender_category = "men" if genders[0] == "M" else "women"
    else:
        gender_category = "mixed"
    
    # Determine age category (most restrictive)
    # Priority: master > senior > j18 > j16
    if "master" in age_categories:
        crew_age_category = "master"
    elif "senior" in age_categories:
        crew_age_category = "senior"
    elif "j18" in age_categories:
        crew_age_category = "j18"
    else:
        crew_age_category = "j16"
    
    # Determine eligible boat types based on crew size
    crew_size = len(crew_members)
    eligible_boat_types = []
    
    if crew_size == 1:
        eligible_boat_types = ["skiff"]
    elif crew_size == 4:
        eligible_boat_types = ["4-"]
    elif crew_size == 5:
        eligible_boat_types = ["4+"]
    elif crew_size == 8 or crew_size == 9:
        eligible_boat_types = ["8+"]
    
    avg_age = sum(ages) / len(ages) if ages else 0
    master_category = get_master_category(avg_age) if crew_age_category == "master" else None
    
    return {
        'crew_size': crew_size,
        'genders': genders,
        'ages': ages,
        'age_categories': age_categories,
        'gender_category': gender_category,
        'age_category': crew_age_category,
        'master_category': master_category,
        'eligible_boat_types': eligible_boat_types,
        'min_age': min(ages) if ages else 0,
        'max_age': max(ages) if ages else 0,
        'avg_age': avg_age
    }


def get_eligible_races(crew_members: List[Dict[str, Any]], available_races: List[Dict[str, Any]]) -> List[Dict[str, Any]]:
    """
    Get races that a crew is eligible for
    
    Args:
        crew_members: List of crew member objects
        available_races: List of available race definitions
    
    Returns:
        List of eligible race objects
    """
    crew_analysis = analyze_crew_composition(crew_members)
    
    if not crew_analysis['eligible_boat_types']:
        return []
    
    eligible_races = []
    
    for race in available_races:
        # Check boat type compatibility
        if race['boat_type'] not in crew_analysis['eligible_boat_types']:
            continue
        
        # Check gender category compatibility
        race_gender = race['gender_category']
        crew_gender = crew_analysis['gender_category']
        
        # Mixed races accept any gender composition
        # Gender-specific races only accept that gender or mixed crews
        if race_gender != "mixed" and race_gender != crew_gender and crew_gender != "mixed":
            continue
        
        # Check age category compatibility
        race_age = race['age_category']
        crew_age = crew_analysis['age_category']
        
        # Age category rules:
        # - J16 can only compete in j16 races
        # - J18 can only compete in j18 races
        # - Senior can compete in senior races
        # - Master can compete in master races with matching category
        if race_age == "j16" and crew_age != "j16":
            continue
        if race_age == "j18" and crew_age != "j18":
            continue
        if race_age == "senior" and crew_age != "senior":
            continue
        if race_age == "master" and crew_age != "master":
            continue
        
        # For master races, check if the race has a specific master category
        if crew_age == "master" and 'master_category' in race:
            crew_master_cat = crew_analysis.get('master_category')
            race_master_cat = race['master_category']
            if crew_master_cat != race_master_cat:
                continue
        
        eligible_races.append(race)
    
    return eligible_races


def validate_race_selection(crew_members: List[Dict[str, Any]], selected_race: Dict[str, Any]) -> Dict[str, Any]:
    """
    Validate that a crew is eligible for a selected race
    
    Args:
        crew_members: List of crew member objects
        selected_race: Selected race object
    
    Returns:
        Dictionary with validation result and details
    """
    crew_analysis = analyze_crew_composition(crew_members)
    
    # Check crew size
    if not crew_analysis['eligible_boat_types']:
        return {
            'valid': False,
            'reason': f"Invalid crew size: {crew_analysis['crew_size']}. Must be 1, 4, 5, 8, or 9 members.",
            'crew_analysis': crew_analysis
        }
    
    # Check boat type
    if selected_race['boat_type'] not in crew_analysis['eligible_boat_types']:
        return {
            'valid': False,
            'reason': f"Boat type '{selected_race['boat_type']}' not compatible with crew size {crew_analysis['crew_size']}",
            'crew_analysis': crew_analysis
        }
    
    # Check gender category
    race_gender = selected_race['gender_category']
    crew_gender = crew_analysis['gender_category']
    
    if race_gender != "mixed" and race_gender != crew_gender and crew_gender != "mixed":
        return {
            'valid': False,
            'reason': f"Gender mismatch: Race requires '{race_gender}' but crew is '{crew_gender}'",
            'crew_analysis': crew_analysis
        }
    
    # Check age category
    race_age = selected_race['age_category']
    crew_age = crew_analysis['age_category']
    
    if race_age == "j16" and crew_age != "j16":
        return {
            'valid': False,
            'reason': f"Age category mismatch: J16 race requires crew members aged 15-16",
            'crew_analysis': crew_analysis
        }
    if race_age == "j18" and crew_age != "j18":
        return {
            'valid': False,
            'reason': f"Age category mismatch: J18 race requires crew members aged 17-18",
            'crew_analysis': crew_analysis
        }
    if race_age == "senior" and crew_age != "senior":
        return {
            'valid': False,
            'reason': f"Age category mismatch: Senior race requires crew members aged 19-26",
            'crew_analysis': crew_analysis
        }
    if race_age == "master" and crew_age != "master":
        return {
            'valid': False,
            'reason': f"Age category mismatch: Master race requires crew members aged 27+",
            'crew_analysis': crew_analysis
        }
    
    return {
        'valid': True,
        'reason': 'Crew is eligible for this race',
        'crew_analysis': crew_analysis
    }


# Boat type requirements for reference
BOAT_TYPE_REQUIREMENTS = {
    'skiff': {
        'crew_size': 1,
        'description': 'Single sculler'
    },
    '4-': {
        'crew_size': 4,
        'description': 'Four without coxswain'
    },
    '4+': {
        'crew_size': 5,
        'description': 'Four with coxswain'
    },
    '8+': {
        'crew_size': [8, 9],  # 8 rowers + optional cox
        'description': 'Eight with coxswain'
    }
}

File: frontend/src/utils/raceEligibility.js
/**
 * Race Eligibility Calculation Engine (Frontend)
 * Determines which races a crew is eligible for based on age and gender composition
 */

import { formatAverageAge } from './formatters.js'

/**
 * Calculate age based on date of birth
 * Age is calculated as the age the person will reach during the current year,
 * regardless of whether their birthday has passed yet.
 * 
 * @param {string} dateOfBirth - Date string in YYYY-MM-DD format
 * @param {Date} referenceDate - Reference date for age calculation (defaults to today)
 * @returns {number} Age the person will reach during the reference year
 */
export function calculateAge(dateOfBirth, referenceDate = new Date()) {
  const birthDate = new Date(dateOfBirth);
  // Age is simply the difference in years - no adjustment for birthday
  const age = referenceDate.getFullYear() - birthDate.getFullYear();
  
  return age;
}

/**
 * Determine age category based on age for display purposes
 * @param {number} age - Age in years
 * @returns {string} Age category (j14, j16, j18, senior, master)
 * 
 * Note: J14 can only be cox. J15 competes in J16 races. J17 competes in J18 races.
 */
export function getAgeCategory(age) {
  if (age === 14) {
    return "j14";
  } else if (age === 15 || age === 16) {
    return "j16";
  } else if (age === 17 || age === 18) {
    return "j18";
  } else if (age < 27) {
    return "senior";
  } else {
    return "master";
  }
}

/**
 * Determine master category letter based on average age of crew
 * @param {number} avgAge - Average age of crew members
 * @returns {string} Master category letter (A, B, C, D, E, F, G, H)
 */
export function getMasterCategory(avgAge) {
  if (avgAge < 36) {
    return "A";  // 27-35
  } else if (avgAge < 43) {
    return "B";  // 36-42
  } else if (avgAge < 50) {
    return "C";  // 43-49
  } else if (avgAge < 55) {
    return "D";  // 50-54
  } else if (avgAge < 60) {
    return "E";  // 55-59
  } else if (avgAge < 65) {
    return "F";  // 60-64
  } else if (avgAge < 70) {
    return "G";  // 65-69
  } else {
    return "H";  // 70+
  }
}

/**
 * Analyze crew composition for race eligibility
 * @param {Array} crewMembers - Array of crew member objects with date_of_birth and gender
 * @returns {Object} Crew composition analysis
 */
export function analyzeCrewComposition(crewMembers) {
  if (!crewMembers || crewMembers.length === 0) {
    return {
      crewSize: 0,
      genders: [],
      ages: [],
      ageCategories: [],
      genderCategory: null,
      ageCategory: null,
      eligibleBoatTypes: []
    };
  }
  
  // Calculate ages and determine categories
  // Note: Age and gender categories are based on ROWERS ONLY (excluding coxswains)
  const ages = [];
  const genders = [];
  const ageCategories = [];
  const rowerAges = [];  // Ages of rowers only (for age category calculation)
  const rowerAgeCategories = [];  // Age categories of rowers only
  const rowerGenders = [];  // Genders of rowers only (for gender category calculation)
  
  crewMembers.forEach(member => {
    const age = calculateAge(member.date_of_birth);
    ages.push(age);
    genders.push(member.gender);
    ageCategories.push(getAgeCategory(age));
    
    // Track rower data separately (exclude coxswains from age and gender category calculations)
    const seatType = member.seat_type || 'rower';
    if (seatType === 'rower') {
      rowerAges.push(age);
      rowerAgeCategories.push(getAgeCategory(age));
      rowerGenders.push(member.gender);
    }
  });
  
  // Determine gender category based on ROWERS ONLY (excluding coxswains)
  // Competition rules:
  // - Women's crews: 100% women rowers
  // - Men's crews: More than 50% men rowers
  // - Mixed-gender crews: At least 1 man AND at least 50% women rowers
  
  // Use rowerGenders instead of genders to exclude coxswains
  let genderCategory;
  let maleCount = 0;
  let femaleCount = 0;
  let malePercentage = 0;
  let femalePercentage = 0;
  
  if (rowerGenders.length > 0) {
    maleCount = rowerGenders.filter(g => g === 'M').length;
    femaleCount = rowerGenders.filter(g => g === 'F').length;
    const totalCount = rowerGenders.length;
    malePercentage = (maleCount / totalCount) * 100;
    femalePercentage = (femaleCount / totalCount) * 100;
    
    if (femaleCount === totalCount) {
      // 100% women rowers
      genderCategory = "women";
    } else if (maleCount > 0 && femalePercentage >= 50) {
      // At least 1 man AND at least 50% women rowers -> Mixed
      genderCategory = "mixed";
    } else if (malePercentage > 50) {
      // More than 50% men rowers -> Men's crew
      genderCategory = "men";
    } else {
      // Edge case: shouldn't happen with valid data
      genderCategory = maleCount >= femaleCount ? "men" : "women";
    }
  } else {
    // Fallback if no rowers (shouldn't happen in valid data)
    maleCount = genders.filter(g => g === 'M').length;
    femaleCount = genders.filter(g => g === 'F').length;
    const totalCount = genders.length;
    malePercentage = totalCount > 0 ? (maleCount / totalCount) * 100 : 0;
    femalePercentage = totalCount > 0 ? (femaleCount / totalCount) * 100 : 0;
    genderCategory = femaleCount === genders.length ? "women" : (maleCount >= femaleCount ? "men" : "women");
  }
  
  // Determine age category (most restrictive) based on ROWERS ONLY
  // Priority: master > senior > j18 > j16
  // Use rowerAgeCategories instead of ageCategories to exclude coxswains
  let crewAgeCategory;
  if (rowerAgeCategories.length > 0) {
    if (rowerAgeCategories.includes("master")) {
      crewAgeCategory = "master";
    } else if (rowerAgeCategories.includes("senior")) {
      crewAgeCategory = "senior";
    } else if (rowerAgeCategories.includes("j18")) {
      crewAgeCategory = "j18";
    } else {
      crewAgeCategory = "j16";
    }
  } else {
    // Fallback if no rowers (shouldn't happen in valid data)
    crewAgeCategory = "j16";
  }
  
  // Determine eligible boat types based on crew size
  // Include both sweep and scull variants where applicable
  const crewSize = crewMembers.length;
  const eligibleBoatTypes = [];
  
  if (crewSize === 1) {
    eligibleBoatTypes.push("skiff");
  } else if (crewSize === 4) {
    eligibleBoatTypes.push("4-");   // Coxless four (sweep)
    eligibleBoatTypes.push("4x-");  // Coxless quad (scull)
  } else if (crewSize === 5) {
    eligibleBoatTypes.push("4+");   // Coxed four (sweep)
    eligibleBoatTypes.push("4x+");  // Coxed quad (scull)
  } else if (crewSize === 8 || crewSize === 9) {
    eligibleBoatTypes.push("8+");   // Eight (sweep)
    eligibleBoatTypes.push("8x+");  // Octuple (scull)
  }
  
  // Calculate average age based on ROWERS ONLY (for master category)
  const avgAge = ages.length > 0 ? ages.reduce((sum, age) => sum + age, 0) / ages.length : 0;
  const rowerAvgAge = rowerAges.length > 0 ? rowerAges.reduce((sum, age) => sum + age, 0) / rowerAges.length : 0;
  const masterCategory = crewAgeCategory === "master" ? getMasterCategory(rowerAvgAge) : null;
  
  return {
    crewSize,
    rowerCount: rowerAges.length,  // Number of rowers (excluding cox)
    genders,
    ages,
    ageCategories,
    genderCategory,
    ageCategory: crewAgeCategory,
    masterCategory,
    eligibleBoatTypes,
    minAge: rowerAges.length > 0 ? Math.min(...rowerAges) : 0,
    maxAge: rowerAges.length > 0 ? Math.max(...rowerAges) : 0,
    avgAge: rowerAvgAge,  // Use rower average age (excluding cox)
    // Gender composition details
    maleCount,
    femaleCount,
    malePercentage: Math.round(malePercentage),
    femalePercentage: Math.round(femalePercentage)
  };
}

/**
 * Get races that a crew is eligible for
 * @param {Array} crewMembers - Array of crew member objects
 * @param {Array} availableRaces - Array of available race definitions
 * @returns {Array} Array of eligible race objects
 */
export function getEligibleRaces(crewMembers, availableRaces) {
  const crewAnalysis = analyzeCrewComposition(crewMembers);
  
  console.log('getEligibleRaces - Crew analysis:', crewAnalysis);
  
  if (crewAnalysis.eligibleBoatTypes.length === 0) {
    console.log('getEligibleRaces - No eligible boat types for crew size:', crewAnalysis.crewSize);
    return [];
  }
  
  const eligibleRaces = [];
  
  availableRaces.forEach(race => {
    // Check boat type compatibility
    if (!crewAnalysis.eligibleBoatTypes.includes(race.boat_type)) {
      return;
    }
    
    // Check gender category compatibility
    const raceGender = race.gender_category;
    const crewGender = crewAnalysis.genderCategory;
    
    // Gender matching rules:
    // - "men" race: only all-men crews
    // - "women" race: only all-women crews
    // - "mixed" race: only crews with both genders (truly mixed)
    if (raceGender !== crewGender) {
      console.log(`getEligibleRaces - Gender mismatch: race=${raceGender}, crew=${crewGender} (${race.race_id})`);
      return;
    }
    
    // Check age category compatibility
    const raceAge = race.age_category;
    const crewAge = crewAnalysis.ageCategory;
    
    // Age category rules:
    // - J16 can only compete in j16 races
    // - J18 can only compete in j18 races
    // - Senior can compete in senior races
    // - Master can compete in master races with matching category
    if (raceAge === "j16" && crewAge !== "j16") {
      return;
    }
    if (raceAge === "j18" && crewAge !== "j18") {
      return;
    }
    if (raceAge === "senior" && crewAge !== "senior") {
      return;
    }
    if (raceAge === "master" && crewAge !== "master") {
      return;
    }
    
    // For master races, check if the race has a specific master category
    // Special rule: Category G can race in F (since no G races exist)
    // All other categories must match exactly
    if (crewAge === "master" && race.master_category) {
      const crewMasterCat = crewAnalysis.masterCategory;
      const raceMasterCat = race.master_category;
      
      console.log(`getEligibleRaces - Master category check: crew=${crewMasterCat}, race=${raceMasterCat} (${race.race_id})`);
      
      // Special exception: G can race in F
      if (crewMasterCat === 'G' && raceMasterCat === 'F') {
        // Allow this combination
      } else if (crewMasterCat !== raceMasterCat) {
        // All other categories must match exactly
        console.log(`getEligibleRaces - Category mismatch, race filtered out`);
        return;
      }
    }
    
    eligibleRaces.push(race);
  });
  
  console.log('getEligibleRaces - Final eligible races:', eligibleRaces.length);
  return eligibleRaces;
}

/**
 * Validate that a crew is eligible for a selected race
 * @param {Array} crewMembers - Array of crew member objects
 * @param {Object} selectedRace - Selected race object
 * @returns {Object} Validation result with details
 */
export function validateRaceSelection(crewMembers, selectedRace) {
  const crewAnalysis = analyzeCrewComposition(crewMembers);
  
  // Check crew size
  if (crewAnalysis.eligibleBoatTypes.length === 0) {
    return {
      valid: false,
      reason: `Invalid crew size: ${crewAnalysis.crewSize}. Must be 1, 4, 5, 8, or 9 members.`,
      crewAnalysis
    };
  }
  
  // Check boat type
  if (!crewAnalysis.eligibleBoatTypes.includes(selectedRace.boat_type)) {
    return {
      valid: false,
      reason: `Boat type '${selectedRace.boat_type}' not compatible with crew size ${crewAnalysis.crewSize}`,
      crewAnalysis
    };
  }
  
  // Check gender category
  const raceGender = selectedRace.gender_category;
  const crewGender = crewAnalysis.genderCategory;
  
  // Strict gender matching: race gender must exactly match crew gender
  if (raceGender !== crewGender) {
    const genderLabels = {
      'men': 'all men',
      'women': 'all women',
      'mixed': 'mixed gender (both men and women)'
    };
    return {
      valid: false,
      reason: `Gender mismatch: Race requires ${genderLabels[raceGender] || raceGender} but crew is ${genderLabels[crewGender] || crewGender}`,
      crewAnalysis
    };
  }
  
  // Check age category
  const raceAge = selectedRace.age_category;
  const crewAge = crewAnalysis.ageCategory;
  
  if (raceAge === "j16" && crewAge !== "j16") {
    return {
      valid: false,
      reason: `Age category mismatch: J16 race requires crew members aged 15-16`,
      crewAnalysis
    };
  }
  if (raceAge === "j18" && crewAge !== "j18") {
    return {
      valid: false,
      reason: `Age category mismatch: J18 race requires crew members aged 17-18`,
      crewAnalysis
    };
  }
  if (raceAge === "senior" && crewAge !== "senior") {
    return {
      valid: false,
      reason: `Age category mismatch: Senior race requires crew members aged 19-26`,
      crewAnalysis
    };
  }
  if (raceAge === "master" && crewAge !== "master") {
    return {
      valid: false,
      reason: `Age category mismatch: Master race requires crew members aged 27+`,
      crewAnalysis
    };
  }
  
  // For master races with specific categories, validate category matching
  // Special rule: Category G can race in F (since no G races exist)
  // All other categories must match exactly
  if (raceAge === "master" && selectedRace.master_category) {
    const crewMasterCat = crewAnalysis.masterCategory;
    const raceMasterCat = selectedRace.master_category;
    
    // Special exception: G can race in F
    if (crewMasterCat === 'G' && raceMasterCat === 'F') {
      // Allow this combination
    } else if (crewMasterCat !== raceMasterCat) {
      return {
        valid: false,
        reason: `Master category mismatch: Category ${crewMasterCat} crew must race in category ${crewMasterCat} (or F if category G)`,
        crewAnalysis
      };
    }
  }
  
  return {
    valid: true,
    reason: 'Crew is eligible for this race',
    crewAnalysis
  };
}

/**
 * Boat type requirements for reference
 */
export const BOAT_TYPE_REQUIREMENTS = {
  'skiff': {
    crewSize: 1,
    description: 'Single sculler'
  },
  '4-': {
    crewSize: 4,
    description: 'Four without coxswain'
  },
  '4+': {
    crewSize: 5,
    description: 'Four with coxswain'
  },
  '8+': {
    crewSize: [8, 9],
    description: 'Eight with coxswain'
  }
};

/**
 * Get human-readable description of crew composition
 * @param {Object} crewAnalysis - Result from analyzeCrewComposition
 * @returns {string} Human-readable description
 */
export function getCrewDescription(crewAnalysis) {
  if (crewAnalysis.crewSize === 0) {
    return "No crew members";
  }
  
  const { rowerCount, genderCategory, ageCategory, masterCategory, avgAge } = crewAnalysis;
  const masterCat = masterCategory ? ` ${masterCategory}` : '';
  return `${rowerCount} ${genderCategory} ${ageCategory}${masterCat} (avg age: ${formatAverageAge(avgAge)})`;
}

/**
 * Get race difficulty/category display
 * @param {Object} race - Race object
 * @returns {string} Display string for race category
 */
export function getRaceDisplay(race) {
  const distance = race.event_type;
  const boat = race.boat_type;
  const gender = race.gender_category;
  const age = race.age_category;
  const masterCat = race.master_category ? ` ${race.master_category}` : '';
  
  return `${distance} ${gender} ${age}${masterCat} ${boat}`;
}
